zio-http/shared/src/main/scala/zio/http/ZClientAspect.scala

```diff
@@ -492,4 +492,104 @@ object ZClientAspect {
           },
         )
     }
+
+  /**
+   * Client aspect that logs details of web request as curl command
+   */
+  final def curlLogger(
+    verbose: Boolean = true,
+    logEffect: String => UIO[Unit] = (m: String) => ZIO.log(m),
+  )(implicit trace: Trace): ZClientAspect[Nothing, Any, Nothing, Body, Nothing, Any, Nothing, Response] =
+    new ZClientAspect[Nothing, Any, Nothing, Body, Nothing, Any, Nothing, Response] {
+
+      def formatCurlCommand(
+        version: Version,
+        method: Method,
+        url: URL,
+        headers: Headers,
+        body: Body,
+        proxy: Option[Proxy],
+      ): String = {
+        val versionOpt = version match {
+          case Version.Default  => Chunk.empty
+          case Version.Http_1_0 => Chunk("--http1.0")
+          case Version.Http_1_1 => Chunk("--http1.1")
+        }
+        val verboseOpt = if (verbose) Chunk("--verbose") else Chunk.empty
+        val requestOpt = method match {
+          case Method.GET          => Chunk("--request GET")
+          case Method.POST         => Chunk("--request POST")
+          case Method.PUT          => Chunk("--request PUT")
+          case Method.DELETE       => Chunk("--request DELETE")
+          case Method.PATCH        => Chunk("--request PATCH")
+          case Method.HEAD         => Chunk("--request HEAD")
+          case Method.OPTIONS      => Chunk("--request OPTIONS")
+          case Method.CONNECT      => Chunk("--request CONNECT")
+          case Method.TRACE        => Chunk("--request TRACE")
+          case Method.CUSTOM(name) => Chunk(s"--request $name")
+          case Method.ANY          => Chunk("--request GET")
+        }
+        val headerOpt  = Chunk.fromIterable(headers.map(h => s"--header '${h.headerName}:${h.renderedValue}'"))
+        val bodyOpt    = body match {
+          case Body.empty => Chunk.empty
+          case body       => {
+            Chunk(s"--data '${body.asString.map(_.replace("'", "'\\''"))}'")
```

Comment:
body.asString returns an effect. Printing an effect is not what you want.
Also, not all bodies can just be transformed to a string.
You need to check `body.isCompleted` to know if a it has data that you are allowed to read.
Else you might consume a stream of the body during logging, and it is not available anymore when we try to send it over the wire.
See the requestLogging client aspect for reference

Reply:
Thanks @987Nabil for spotting those issues! 🙏
The body  is now only converted to a string if `body.isCompleted` is `true`. Also `body.asChunk` is now correctly evaluated to get the result of the effect.